### PR TITLE
basic prometheus metrics

### PR DESCRIPTION
**Pull request by kikht**
This patch exposes some basic gobgp metrics in prometheus format. It reuses existing http endpoint that is used for pprof. For now only very basic metrics are implemented:
1. Per-neighbor bgp message counters. 
2. Number of recieved/accepted routes for each neighbor and each route family. 
3. GRPC api counters. 

Issue: #1507

### CODE REVIEW

**Comment by fujita** [comment]
Why not using the gRPC API like this?
https://github.com/greenpau/gobgp_exporter

**Comment by kikht** [comment]
> Why not using the gRPC API like this? https://github.com/greenpau/gobgp_exporter

Mostly for the grpc api call counters. While it is possible to collect them and extend api to expose them, it is much easier to do it directly in prometheus format. Also having metrics in gobgpd is slightly more convinient from deployment perspective - we have only single self-contained daemon binary.

**Comment by kikht** [comment]
@fujita gentle ping. Is this approach feasible or we should pass all necessary metrics through gobgp_exporter?

**Comment by pavel-odintsov** [comment]
Really nice patch, we added native support for Prometheus while ago and customers love it.

File: cmd/gobgpd/main.go [hunk 1]
```diff
@@ -88,7 +91,14 @@ func main() {
 		runtime.GOMAXPROCS(opts.CPUs)
 	}
 
+	metricRegistry := prometheus.NewRegistry()
+
 	if !opts.PProfDisable {
```
Comment (mrueg) [inline cmd/gobgpd/main.go:96]: Should this line be changed? Otherwise if you disable pprof, you will also disable metrics, right?

**Comment by fujita** [comment]
Monitoring software comes and goes. So I prefer the feature to be implemented in a way that it would be replaced with a different software easily. If people prefer native prometheus, it's fine by me, however, I don't want the prometheus support to play with GoBGP internal directly. Instead, I prefer it to play with the gRPC APIs.

**Comment by kikht** [comment]
Sure, I can take less invasive approach and use only public api to get necessary information. Please, see the latest commits. Now all metrics stuff is moved to another package and uses public ListPeer method like grpc server does. This way we only need to add initialization stuff to gobgpd main and do not otherwise mess with its internals.

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
+package metrics
```
Comment (fujita) [inline pkg/metrics/metrics.go:1]: `internal/pkg/metrics` is more appropriate place?

**Comment by fujita** [comment]
> Now all metrics stuff is moved to another package and uses public ListPeer method like grpc server does. This way we only need to add initialization stuff to gobgpd main and do not otherwise mess with its internals.

Thanks, looks good. But I have only one comment, why this is placed at `/pkg/metrics/` instead of `/internal/pkg/metrics/`?

**Comment by fujita** [comment]
Thanks, pushed.
I wasn't sure about supporting prometheus natively last time but let's see.

**Comment by kochkarev** [comment]
@fujita @kikht Hi! In my opinion it would be convenient to have this package in `/pkg/metrics/` instead of `/internal/pkg/metrics/`. GoBGP server can be imported and used while metrics package has to be copied :(

**Pull request closed without merging**